### PR TITLE
Optimizing communication with NCCL and RCCL.

### PR DESCRIPTION
**Pull request by kabicm**
This PR enables COSMA to take advantage of fast GPU-to-GPU interconnects like NVLink, to efficiently utilize modern Multi-GPU Systems. This is achieved in 2 ways:
- **Using `NCCL/RCCL` Libraries:** by specifying `-DCOSMA_WITH_NCCL=ON` (for NVIDIA GPUs) or `-DCOSMA_WITH_RCCL=ON` (for AMD GPUs) cmake options.
- **Using GPU-aware MPI:** by specifying `-DCOSMA_WITH_GPU_AWARE_MPI=ON` cmake option, as proposed [here](https://github.com/eth-cscs/COSMA/pull/101#issuecomment-1126514781).
See [README](https://github.com/eth-cscs/COSMA/blob/master/README.md) and [INSTALL](https://github.com/eth-cscs/COSMA/blob/master/INSTALL.md) for more info on how to build.

In addition, the following performance improvemets have been made:
- **Improved Caching:** 
    - all nccl buffers, MPI comms, nccl comms are cached and reused when appropriate.
    - all device memory is cached and reused.
- **Reduced Data Trasfers:** the GPU backend of COSMA called [Tiled-MM](https://github.com/eth-cscs/Tiled-MM) is extended to offer the possibility to the user to leave the resulting matrix C on the GPU. In that case, there is no need to trasfer matrix C from device to host, which not only reduces the communication, but also speeds up the whole cpu->gpu pipeline as no additional synchronizations are needed. Furthermore, reduce_scatter operation does not have to wait for C to be transfered back to host but is immediately invoked with GPU pointers, thus utilizing fast inter-gpu links. This way, there is no unnecessary data transfers between cpu<->gpu.
- **All collectives updated:** both `all-gather` and `reduce-scatter` collectives are improved.
- **Reduced Data Reshuffling:** avoids double reshuffling of data, i.e. the data from NCCL/RCCL GPU buffers is immediately copied in the right layout, without additional reshuffling.
- **Works for variable blocks:** NCCL/RCCL' reduce_scatter operation assumes that all the blocks are of the same size and is hence not completely equivalent to `MPI_Reduce_scatterv` which we previously used. We padded all the blocks to be able to overcome this issue.
- **Portability:** Supports both NVIDIA and AMD GPUs.

Therefore, this fixes the limitations of https://github.com/eth-cscs/COSMA/pull/101 and brings above-mentioned improvements.

Thanks to @alazzaro and @gsitaram for their great feedback and contribution to this PR!

### CODE REVIEW

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try

**Comment by haampie** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
@haampie do you know why it fails with MKL but succeeds with OPENBLAS? It complain about OpenMP. COSMA doesnt use OpenMP as a direct dependency. It's only used by COSTA and is then passed to COSMA as an indirect dependency.

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by kabicm** [comment]
bors try

**Comment by kabicm** [comment]
bors try-

**Comment by alazzaro** [comment]
What's the timeline of this PR?

File: src/cosma/gpu/utils.hpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    // the default (NULL) stream
+    template <typename T>
+    void copy_to_host_async(const T* from, T* to, size_t n, runtime_api::StreamType stream=NULL) {
+        auto status = runtime_api::memcpy_async(to, from, n * sizeof(T),
+                runtime_api::flag::MemcpyDeviceToHost, stream);
+        check_runtime_status(status);
+    }
+
+    template <typename Scalar>
+    void nccl_reduce(
```
Comment (BORIS-ADMIN) [inline src/cosma/gpu/utils.hpp:65]: @gsitaram

File: src/cosma/gpu/utils.hpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    // the default (NULL) stream
+    template <typename T>
+    void copy_to_host_async(const T* from, T* to, size_t n, runtime_api::StreamType stream=NULL) {
+        auto status = runtime_api::memcpy_async(to, from, n * sizeof(T),
+                                                runtime_api::flag::MemcpyDeviceToHost, stream);
+        check_runtime_status(status);
+    }
+
+    template <typename Scalar>
+    void nccl_reduce(
```
Comment (kabicm) [inline src/cosma/gpu/utils.hpp:65]: @gsitaram

**Comment by alazzaro** [comment]
OK, cross-referencing what I wrote at https://github.com/eth-cscs/COSMA/pull/101#issuecomment-1126514781, I did a comparison of this branch (NCCL) versus COSMA v2.5.1, with HIP enabled (no RCCL), i.e. with the following cmake flags:

```
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=${PWD}/install -DCOSMA_BLAS=ROCM -DCOSMA_SCALAPACK=CRAY_LIBSCI -DCOSMA_WITH_TESTS=NO -DCOSMA_WITH_BENCHMARKS=NO -DCMAKE_CXX_COMPILER=CC -DCOSMA_WITH_APPS=YES -DCOSMA_WITH_PROFILING=ON -DBUILD_SHARED_LIBS=ON ..
```

I'm using the H2O-128 RPA CP2K benchmark. Then I set the following env variables:

```
export COSMA_GPU_MAX_TILE_M=30000
export COSMA_GPU_MAX_TILE_N=30000
export COSMA_GPU_MAX_TILE_K=30000
export COSMA_DIM_THRESHOLD=10000 # force only RPA calls to run via COSMA                                                                                                                                                                                                                                    
export COSMA_CPU_MEMORY_ALIGNMENT=256
```

It turns out that the NCCL branch is slower than v2.5.1. I'm reporting here the outputs of the COSMA profiler for a multiplication (more or less they are all the same) when running on the same set of nodes:

**NCCL branch**
```
p_ REGION                     CALLS      THREAD        WALL       %
_p_ total                          -       4.325       4.325   100.0
_p_   multiply                     -       3.063       3.063    70.8
_p_     computation                4       2.089       2.089    48.3
_p_     communication              -       0.975       0.975    22.5
_p_       reduce                   4       0.972       0.972    22.5
_p_       other                    8       0.003       0.003     0.1
_p_     other                     30       0.000       0.000     0.0
_p_   transform                    -       1.261       1.261    29.2
_p_     localblocks                1       1.261       1.261    29.2
_p_     init                       1       0.000       0.000     0.0
_p_     decompose                  4       0.000       0.000     0.0
_p_     packing                    1       0.000       0.000     0.0
_p_     waitall                    1       0.000       0.000     0.0
_p_     isend                      1       0.000       0.000     0.0
_p_     irecv                      1       0.000       0.000     0.0
_p_     waitany                    0       0.000       0.000     0.0
_p_     unpacking                  0       0.000       0.000     0.0
_p_     reordering                 -       0.000       0.000     0.0
_p_       comm                     0       0.000       0.000     0.0
_p_   preprocessing                -       0.001       0.001     0.0
_p_     matrices                   -       0.001       0.001     0.0
_p_       mapper                   -       0.000       0.000     0.0
_p_         sizes                  3       0.000       0.000     0.0
_p_         coordinates            3       0.000       0.000     0.0
_p_       buffer                   3       0.000       0.000     0.0
_p_       layout                   3       0.000       0.000     0.0
_p_     allocation                 2       0.000       0.000     0.0
_p_     communicators              0       0.000       0.000     0.0
_p_   strategy                     1       0.000       0.000     0.0
_p_   init                         2       0.000       0.000     0.0

```

**v2.5.1**
```
_p_ REGION                     CALLS      THREAD        WALL       %
_p_ total                          -       3.881       3.881   100.0
_p_   multiply                     -       2.593       2.593    66.8
_p_     computation                4       1.815       1.815    46.8
_p_     communication              -       0.778       0.778    20.0
_p_       reduce                   4       0.776       0.776    20.0
_p_       other                    8       0.002       0.002     0.1
_p_     other                     30       0.000       0.000     0.0
_p_   transform                    -       1.245       1.245    32.1
_p_     localblocks                1       1.245       1.245    32.1
_p_     init                       1       0.000       0.000     0.0
_p_     decompose                  4       0.000       0.000     0.0
_p_     irecv                      1       0.000       0.000     0.0
_p_     isend                      1       0.000       0.000     0.0
_p_     packing                    1       0.000       0.000     0.0
_p_     waitall                    1       0.000       0.000     0.0
_p_     waitany                    0       0.000       0.000     0.0
_p_     unpacking                  0       0.000       0.000     0.0
_p_     reordering                 -       0.000       0.000     0.0
_p_       comm                     0       0.000       0.000     0.0
_p_   preprocessing                -       0.043       0.043     1.1
_p_     communicators              1       0.042       0.042     1.1
_p_     matrices                   -       0.001       0.001     0.0
_p_       mapper                   -       0.000       0.000     0.0
_p_         sizes                  3       0.000       0.000     0.0
_p_         coordinates            3       0.000       0.000     0.0
_p_       buffer                   3       0.000       0.000     0.0
_p_       layout                   3       0.000       0.000     0.0
_p_     allocation                 2       0.000       0.000     0.0
_p_   strategy                     1       0.000       0.000     0.0
_p_   init                         2       0.000       0.000     0.0
```

As you can see, both `computation` and `communication` are slower with the NCCL branch, but I'm not sure why... In the new code, I see printed `alignment = 256` tons to times (coming from [this line](https://github.com/eth-cscs/COSMA/blob/d846c50fc5efe8ba1d05afff5dac6b8d44be48b9/src/cosma/memory_pool.cpp#L131). Commenting it, it doesn't solve the slowness. I tried to remove the `export COSMA_CPU_MEMORY_ALIGNMENT=256` and it doesn't help.

**Comment by kabicm** [comment]
Thanks @alazzaro for this feedback. The main difference between these two versions is the following:
- v2.5.1: for local multiplication, the matrices are tiled and then sent tile by tile to gpus for multiplication, multiplied, and sent back, all in the pipelined fashion. After the result is on the host, MPI_Reduce_scatter is invoked.
- nccl: for local multiplication, the full matrices are sent to gpus, multiplied, reduce-scattered with nccl and then sent back. 

I believe the lack of tiling might make nccl branch slower. However, in your case, the tile sizes are already pretty large, so this might be equivalent to having no tiiling at all.

I will have to test this.

Can you tell me on how many ranks was this run?

**Comment by alazzaro** [comment]
> Can you tell me on how many ranks was this run?

128 ranks. 
Note, nccl (or rccl) is not the first solution I will think about if I have an efficient G2G MPI installed on the system. 
Took me a while to install RCCL (with AMD people's help) to get full performance, while G2G MPI is there for free...

**Comment by kabicm** [comment]
@alazzaro I am now adding the tiling logic to this branch.

If gpu-aware MPI is working out of the box with comparable performance (?), we would be interested in using it. 

Still, the changes made in this PR will be necessary for the gpu-aware MPI version as well.

**Comment by alazzaro** [comment]
Well, the G2G requires to call the MPI function with GPU pointers data, then you rely on the MPI vendor for the performance. For RCCL or NCCL, there are GPU specific and need special plugins to enable the network.
I can give you a bit of numbers for the code I have reported at https://github.com/eth-cscs/COSMA/pull/101#issuecomment-1126514781. I put a set of timers around each operation (malloc, h2d, MPI, d2h, free) and those are the timings (COSMA v2.5.1 version):

```
p_ REGION                     CALLS      THREAD        WALL       %
_p_ total                          -       3.881       3.881   100.0
_p_   multiply                     -       2.593       2.593    66.8
_p_     computation                4       1.815       1.815    46.8
_p_     communication              -       0.778       0.778    20.0
_p_       reduce                   -       0.776       0.776    20.0
_p_         mpi                    4       0.580       0.580    14.9
_p_         h2d                    4       0.190       0.190     4.9
_p_         d2h                    4       0.005       0.005     0.1
_p_         malloc                 4       0.001       0.001     0.0
_p_         free                   4       0.000       0.000     0.0
_p_       other                    8       0.002       0.002     0.1
_p_     other                     30       0.000       0.000     0.0
_p_   transform                    -       1.245       1.245    32.1
_p_     localblocks                1       1.245       1.245    32.1
_p_     init                       1       0.000       0.000     0.0
_p_     decompose                  4       0.000       0.000     0.0
_p_     irecv                      1       0.000       0.000     0.0
_p_     isend                      1       0.000       0.000     0.0
_p_     packing                    1       0.000       0.000     0.0
_p_     waitall                    1       0.000       0.000     0.0
_p_     waitany                    0       0.000       0.000     0.0
_p_     unpacking                  0       0.000       0.000     0.0
_p_     reordering                 -       0.000       0.000     0.0
_p_       comm                     0       0.000       0.000     0.0
_p_   preprocessing                -       0.043       0.043     1.1
_p_     communicators              1       0.042       0.042     1.1
_p_     matrices                   -       0.001       0.001     0.0
_p_       mapper                   -       0.000       0.000     0.0
_p_         sizes                  3       0.000       0.000     0.0
_p_         coordinates            3       0.000       0.000     0.0
_p_       buffer                   3       0.000       0.000     0.0
_p_       layout                   3       0.000       0.000     0.0
_p_     allocation                 2       0.000       0.000     0.0
_p_   strategy                     1       0.000       0.000     0.0
_p_   init                         2       0.000       0.000     0.0

```

As you can see, the `h2d` takes 0.2 seconds, and I think we can avoid it by keeping alive the buffers on the GPU. However, the NCCL branch is 0.5 seconds slower, so it will not pay off... If you are telling that we cannot improve on top of the current v2.5.1, then I will stick with that version + my change to enable G2G...

**Comment by kabicm** [comment]
This PR enables COSMA to take advantage of fast GPU-to-GPU interconnects like NVLink, to efficiently utilize modern Multi-GPU Systems. This is achieved in 2 ways:
- **Using `NCCL/RCCL` Libraries:** by specifying `-DCOSMA_WITH_NCCL=ON` (for NVIDIA GPUs) or `-DCOSMA_WITH_RCCL=ON` (for AMD GPUs) cmake options.
- **Using GPU-aware MPI:** by specifying `-DCOSMA_WITH_GPU_AWARE_MPI=ON` cmake option, as proposed [here](https://github.com/eth-cscs/COSMA/pull/101#issuecomment-1126514781).
See [README](https://github.com/eth-cscs/COSMA/blob/master/README.md) and [INSTALL](https://github.com/eth-cscs/COSMA/blob/master/INSTALL.md) for more info on how to build.

In addition, the following performance improvemets have been made:
- **Improved Caching:** 
    - all nccl buffers, MPI comms, nccl comms are cached and reused when appropriate.
    - all device memory is cached and reused.
- **Reduced Data Trasfers:** the GPU backend of COSMA called [Tiled-MM](https://github.com/eth-cscs/Tiled-MM) is extended to offer the possibility to the user to leave the resulting matrix C on the GPU. In that case, there is no need to trasfer matrix C from device to host, which not only reduces the communication, but also speeds up the whole cpu->gpu pipeline as no additional synchronizations are needed. Furthermore, reduce_scatter operation does not have to wait for C to be transfered back to host but is immediately invoked with GPU pointers, thus utilizing fast inter-gpu links. This way, there is no unnecessary data transfers between cpu<->gpu.
- **All collectives updated:** both `all-gather` and `reduce-scatter` collectives are improved.
- **Reduced Data Reshuffling:** avoids double reshuffling of data, i.e. the data from NCCL/RCCL GPU buffers is immediately copied in the right layout, without additional reshuffling.
- **Works for variable blocks:** NCCL/RCCL' reduce_scatter operation assumes that all the blocks are of the same size and is hence not completely equivalent to `MPI_Reduce_scatterv` which we previously used. We padded all the blocks to be able to overcome this issue.
- **Portability:** Supports both NVIDIA and AMD GPUs.

Therefore, this fixes the limitations of https://github.com/eth-cscs/COSMA/pull/101 and brings above-mentioned improvements.

Thanks to @alazzaro and @gsitaram for their great feedback and contribution to this PR!

**Comment by alazzaro** [comment]
@kabicm I did a test with the new COSMA and I can confirm that I get some quite nice speed-up with the G2G MPI option enabled for the CP2K RPA benchmark. Note that I'm using (my default setting to run the benchmark):

```
    export COSMA_GPU_MAX_TILE_M=30000
    export COSMA_GPU_MAX_TILE_N=30000
    export COSMA_GPU_MAX_TILE_K=30000

    export COSMA_CPU_MEMORY_ALIGNMENT=256

    export COSMA_DIM_THRESHOLD=10000 # force only RPA calls to run via COSMA
```

Now, I tried to relax some of the variables. 
If I remove the last variable, so all pdgemm's (including small ones) will go to COSMA, I get an error:

```
Backtrace for this error:

Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

gdb) where
#0  0x00007f13e759345d in cosma::Interval::locate_in_subinterval (this=0x7ffd99185980, divisor=1072693248, elem=28) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/interval.cpp:66
#1  0x00007f13e7600c17 in cosma::gpu::gpu_aware_mpi_copy<double> (ctx=0x1b62d270, P=..., in=0x1bf44500, out=0x1c0a8500, reshuffle_buffer=0x0, size_before=..., total_before=..., total_after=58368, step=1) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/gpu/gpu_aware_mpi_utils.cpp:28
#2  0x00007f13e75c260f in cosma::parallel<double> (ctx=0x1b62d270, matrixA=..., matrixB=..., matrixC=..., m=..., n=..., k=..., P=..., step=1, strategy=..., comm=0x1b2276f0, alpha=1, beta=0) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/multiply.cpp:776
#3  0x00007f13e75bf88c in cosma::multiply<double> (ctx=0x1b62d270, matrixA=..., matrixB=..., matrixC=..., m=..., n=..., k=..., P=..., step=1, strategy=..., comm=0x1b2276f0, alpha=1, beta=0) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/multiply.cpp:411
#4  0x00007f13e75c26e3 in cosma::parallel<double> (ctx=0x1b62d270, matrixA=..., matrixB=..., matrixC=..., m=..., n=..., k=..., P=..., step=0, strategy=..., comm=0x1b2276f0, alpha=1, beta=0) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/multiply.cpp:881
#5  0x00007f13e75bf88c in cosma::multiply<double> (ctx=0x1b62d270, matrixA=..., matrixB=..., matrixC=..., m=..., n=..., k=..., P=..., step=0, strategy=..., comm=0x1b2276f0, alpha=1, beta=0) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/multiply.cpp:411
#6  0x00007f13e75bc537 in cosma::multiply<double> (ctx=0x1b62d270, matrixA=..., matrixB=..., matrixC=..., strategy=..., comm=-1006632895, alpha=1, beta=0) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/multiply.cpp:283
#7  0x00007f13e75bcde5 in cosma::multiply<double> (matrixA=..., matrixB=..., matrixC=..., strategy=..., comm=-1006632895, alpha=1, beta=0) at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/multiply.cpp:228
#8  0x00007f13e788736e in cosma::pxgemm<double> (transa=84 'T', transb=78 'N', m=512, n=512, k=7296, alpha=1, a=0x20139060, ia=1, ja=1, desca=0x16e9df50, b=0x20139060, ib=1, jb=1, descb=0x16e9df50, beta=0, c=0x1d8faa90, ic=1, jc=1, descc=0x200f47a0)
    at /home/users/alazzaro/cp2k/libs/COSMA-marko/src/cosma/cosma_pxgemm.cpp:350

```

I assume the problem is that there is a reshuffling of the operations for small multiplications (a small number of ranks involved?)
Maybe you may want to investigate more, let me know if you need more info.

I will also run some more tests at scale (in ~ a month timescale). Thanks!

**Pull request merged**